### PR TITLE
Allow orthoManagers to be configured using nested parameter list

### PR DESCRIPTION
**Pull request by jjellio**
Update OrthOmanagers and Belos Pseudo Block GMRES to allow parameterlist configurations

API works as:
Solver Parameter List:
-- Orthogonalization : DGKS | ICGS | IMGS | TSQR
-- DGKS (sublist)
---- ...
-- ICGS (sublist)
---- ...

Ortho parameters are valid if they are acceptable as setParameters()
This allows a user to fully configure the orthogonalizer.

WARNING:
It is not clear how to handle the "Orthogonalization Constant" (kappa) that
is baked into GMRES. Currently, this constant's default is enforced at the GMRES
level.

Additions
1) allows the use of a sublist to control the orthomanager
   The solver takes a string parameter named "Orthogonalization"
   "Orthogonalization" accepts the same values it always has: DGKS, ICGS, IMGS, TSQR

   The solver constructs its default parameter list by calling the OrthoManager's new
   static function, getDefaultParameters. A sublist is added to the solver's paramater
   list named DGKS, ICGS, IMGS, and TSQR.

   An orthoManager is instantiated based off the Orthogonalization name, and
   the sublist if present, is passed to the constructor.

   This required the addition of:
   static template<const ParameterList> getDefaultParameters ()

   And the modification of:
   getValidParameters in the solver
   setParameters in the solver

   Note:
   The current implementations convolute getValidParameters with getDefaultParameters.
   A small change would be to have getValidParameters call getDefaultParameters.
   The former uses the defaults, and performs validation. The latter provides defaults
   without needed the class to be instantiated.

   The motivation for getDefaultParameters is that we can validate parameters without instantiating
   an orthoManager. E.g., we can query acceptable parameters without instantiation, and then
   validate against this list at the solver level.

   Orthogonalization sublists are validated. This will ensure that if a user provides a parameter,
   it will either be used or an error generated. This resolves the current problem, where apps
   and examples set the number of Ortho sweeps, but this does nothing.

2) Adjusted the labels used for Teuchos timers inside the orthoManagers.
   Labels now take the form: label_ + kernel_name + [ortho_passes] + detailed label
   E.g., Belos: ICGS[1]: Ortho (Inner Product)

   This addresses a performance concern, where users experience slow performance
   because the default ortho kernel is robust  numerically, but has slow performance relative
   to classical gram schmidt. This is a compromise between changing the default, and ignoring
   this user issue. This way, users are informed as best as possible, and performance

@mhoemmen @hkthorn

### CODE REVIEW

**Comment by jjellio** [comment]
@mhoemmen and @hkthorn 

I would like to discuss this pull request.  I currently have tested with PseudoBlockGMRES, and I would need to add the code to make it work for other GMRES implementations. As well as other Belos solvers that use ortho managers.

There are two disjoint features: timer labels and ortho parameters. I can split them, but both are responses to the Trilinos Performance work that is ongoing.

**Comment by mhoemmen** [comment]
@jjellio I'm curious -- why did you choose to make the new sublist have the same name as the existing parameter "Orthogonalization"?  Would it make sense to add a new parameter, say "Orthogonalization Parameters", for the sublist?

File: packages/belos/src/BelosPseudoBlockGmresSolMgr.hpp [hunk 1]
```diff
@@ -1167,11 +1229,55 @@ PseudoBlockGmresSolMgr<ScalarType,MV,OP>::getValidParameters() const
...
+      RCP<ortho_type> ortho_tmp = rcp (new ortho_type ());
+
+      RCP<ParameterList> orthoParams = sublist(pl, "TSQR");
+      orthoParams->setParameters( *(ortho_tmp->getValidParameters()) );
+    }
+    #endif // HAVE_BELOS_TSQR
+    //-----------------------------------------------------------------------
+
+    // legacy parameters
+    Teuchos::RCP<Teuchos::ParameterList> orthoPl = Teuchos::sublist(pl, "Orthogonalziation");
```
Comment (mhoemmen) [inline packages/belos/src/BelosPseudoBlockGmresSolMgr.hpp]: "Orthogonalization" is misspelled here.

Comment (jjellio) [inline packages/belos/src/BelosPseudoBlockGmresSolMgr.hpp]: This is cruft. And why you saw that odd output.  I hadn't generated that list of 'valid parameters' before I emailed them.  This was from a first shot at this. The motivation  was to have a single sublist for the ortho parameters ... but that is foolish, each Ortho kernel has its own set of parameters. It also requires a 'Name' parameter somewhere, e.g., added to each OrthoManager's valdParameters, that way you know which orthogonalization to instantiated.

Instead, I deleted *most* of that... and went with a more natural OO approach.
The solver keeps the original Orthogonalization string (DGKS, ...), and that is used to look up an optional sublist. This lets you append all valid ortho manager paramlists as sublists using their default names (ICGS, DGKS, etc..). Thus creating a valid parameter list in the solver that is both comprehensive and allows validation.

I'll need to clean up a few more things as well, but I wanted to get criticism on the approach before I spend more time on it (pushing to other solvers).

**Comment by hkthorn** [comment]
@mhoemmen @jjellio Ok, so I haven't looked at the pull request, my head is in something different right now.  However, in looking at Belos over the last couple weeks I have noticed there are already two different ways that orthogonalization is managed through parameter lists: directly in each solver manager and through the OrthoManagerFactory (used by GCRO-DR).  I've also been thinking about how changes like these will affect the end user.  So, if the management of orthogonalization in the Belos solvers is to be changed, I think there are some issues to consider to improve readability, maintainability, and usability:

1)  Move to one approach to manage orthogonalization for all solver managers in Belos.  We have two ways now, I don't want to have three.  I can barely remember the first.
2)  This approach must also support delta-semantics.  The parameters passed into a Belos solver object are expected to only modify the current settings, not reset all settings.  This has been broken by a couple newer solvers and I would like to preserve the approach in those that do support it.
3)  The code added to each solver manager, or orthogonalization manager, should be short and readable.  Belos gets a lot of comments about ease of understanding, the parameter list management is just one portion of the code that is replicated many (MANY times, Mark has mentioned this before to me :) ).  Reduction in the amount of parameter processing performed by each solver manager is preferred.
4)  How does this look, in terms of parameters, to application users?  In seeing the XML parameter lists used by Albany, Charon, etc. there is an extensive list of parameter options they try to keep track of in their vertically integrated software stack.  Many of them they don't necessarily care about.  While orthogonalization is useful to Belos, it is of little interest to application codes.  I'd be hesitant to expand the parameter list further and give them more parameters they don't need to know about.

I hope to get to looking at the pull request tomorrow and I'll let you know more thoughts then.  Thanks!

**Comment by jjellio** [comment]
The short:
Perhaps I should add the changes to use the ortho factory, clean up the now unnecessary code. Look at GCRO-DR, And then report back.

@hkthorn 
I agree 100% with 1-3. The details of 2 are tricky, and why I am requesting comments/guidence  prior to simply pushing to develop.

4. I will disagree on.  The app I have been working with wants to explore algorithmic and system performance trade offs. Currently, the only option is to choose between GMRES + DGKS(2) or GMRES+ICGS(2), (IMGS, TSQR aswell), but the number of ortho passes if effectively hard coded.
In this particular app's case, they are looking at things like nonlinear convergence with the correct Jacobian versus a cheaper approximation.  Also, they are not using a large subspace. It is very possible that numerical stability in ortho is completely overshadowed by the cheap Jacobian. In this case, we could see great performance gains by being able to configure the orthogonalization kernel.

Also, this is not adding any new parameters. It is simply exposing the ones that already exist and are used. If you choose not to set a sublit for Ortho, then you get what you always had. GCRO-DR may has solved this correctly.

GCRO-DR added an "Orthogonalization Parameters" sublist.

```
          <Parameter docString='The type of orthogonalization to use.  Valid options: "DGKS", "ICGS", "IMGS", "TSQR", or "Simple"' id="78" isDefault="false" isUsed="true" name="Orthogonalization" type="string" value="DGKS"/>
          <ParameterList id="83" name="Orthogonalization Parameters">
            <Parameter docString='Maximum number of orthogonalization passes (includes the first).  Default is 2, since "twice is enough" for Krylov methods.' id="79" isDefault="false" isUsed="true" name="maxNumOrthogPasses" type="int" value="2"/>
            <Parameter docString="Block reorthogonalization threshhold." id="80" isDefault="false" isUsed="true" name="blkTol" type="double" value="1.49011611938476562e-07"/>
            <Parameter docString="(Non-block) reorthogonalization threshold." id="81" isDefault="false" isUsed="true" name="depTol" type="double" value="7.07106781186547462e-01"/>
            <Parameter docString="Singular block detection threshold." id="82" isDefault="false" isUsed="true" name="singTol" type="double" value="2.22044604925031308e-15"/>
          </ParameterList>
          <Parameter docString='When using DGKS orthogonalization: the "depTol" constant, used to determine whether another step of classical Gram-Schmidt is necessary.  Otherwise ignored.' id="84" isDefault="false" isUsed="true" name="Orthogonalization Constant" type="double" value="0.00000000000000000e+00"/>
        </ParameterList>
```

The problem, is that GCRO-DR is now embedding this OrthoManager knowledge, into it's configuration information. The names used, are relevant to DGKS, but may be meaningless to a different Ortho Kernel.

My Idea:
The sublist name is named by the ortho type.
This sublist can then be *validated* by the orthoManager, and it's defaults also pulled from the ortho manager.
In this way, the solver (or OrthoManagerFactory) can correctly create that parameter list with the correct fields given the orthogonalization type. The solve needs no knowledge of the ortho details, just a key "DGKS", "ICGS", and using that key (or the default) it simply asks the OrthoFactory or OrthoManagers for valid parameters. Ortho knowledge stays in the OrthoManagers, and solver's complexity is reduced.

The valid parameters for GMRES would then correctly list the possible ortho pamareters, while the default parameters would look something like:

Orthogonalization Type: DGKS
DGKS {
 depTol : -1 (the current GMRES default)
 ... other DGKS parameters
}

I think this is superior and more maintainable than what GCRO-DR is doing.

Apologies for the length, this could probably be explained on a white board much faster :)

File: packages/belos/src/BelosPseudoBlockGmresSolMgr.hpp [hunk 2]
```diff
@@ -735,9 +737,28 @@ setParameters (const Teuchos::RCP<Teuchos::ParameterList>& params)
     }
   }
 
+  // Optional sublist for OrthoManager parameters
+  RCP<ParameterList> orthoParams;
+  std::string tempOrthoType;
+  // the sublist for ortho options is new (2017).
+  // the intent is that the string parameter 'Orthogonalization'
+  // can optionally define a sublist.
+  const bool haveOrthoParamName = params->isParameter ("Orthogonalization");
```
Comment (hkthorn) [inline packages/belos/src/BelosPseudoBlockGmresSolMgr.hpp]: I'm not sure why this logic, the ParameterList get() method will return the default if that parameter doesn't exist.

File: packages/belos/src/BelosPseudoBlockGmresSolMgr.hpp [hunk 3]
```diff
@@ -758,32 +779,52 @@ setParameters (const Teuchos::RCP<Teuchos::ParameterList>& params)
 
     if (tempOrthoType != orthoType_) {
       orthoType_ = tempOrthoType;
+
       params_->set("Orthogonalization", orthoType_);
+
+      if (haveOrthoSublist) {
+        params_->set("OrthoManager", *orthoParams);
```
Comment (hkthorn) [inline packages/belos/src/BelosPseudoBlockGmresSolMgr.hpp]: The list that is returned by getCurrentParameters() needs to be readable if it is given to setParameters().  The use of "OrthoManager" as the sublist name doesn't seem to be consistent with how the parameter list is processed by the solver manager, as it is looking for a sublist having the name of the orthogonalization type.

File: packages/belos/src/BelosPseudoBlockGmresSolMgr.hpp [hunk 4]
```diff
@@ -758,32 +779,52 @@ setParameters (const Teuchos::RCP<Teuchos::ParameterList>& params)
...
+      }
+
       // Create orthogonalization manager
       if (orthoType_ == "DGKS") {
         typedef DGKSOrthoManager<ScalarType, MV, OP> ortho_type;
         if (orthoKappa_ <= 0) {
-          ortho_ = rcp (new ortho_type (label_));
```
Comment (hkthorn) [inline packages/belos/src/BelosPseudoBlockGmresSolMgr.hpp]: Ok, all of this additional logic for processing the orthogonalization type and parameters I think should be relegated to the orthogonalization manager.  This logic block will be repetitive among all iterative solvers that require orthogonalization.  Furthermore, I do not know of anyone who actually changes the "Orthogonalization Constant" from DGKS.  So, I would be ok just removing that logic from all solver managers and moving the parameter into the sublist.

**Comment by hkthorn** [comment]
@jjellio @mhoemmen I have commented on the changes in the pull request.

File: packages/belos/src/BelosPseudoBlockGmresSolMgr.hpp (hunk 2 above)
Comment (mhoemmen) [inline packages/belos/src/BelosPseudoBlockGmresSolMgr.hpp]: @hkthorn Two-argument ParameterList::get() actually adds the parameter to the list with the default value if it's not already there, right?

**Pull request closed without merging**